src/aspire/source/coordinates.py

```diff
@@ -364,10 +364,18 @@ def _extract_ctf(self, df):
             "_rlnMicrographPixelSize",
         ]
 
+        n = len(df["_rlnVoltage"])
         # get unique ctfs from the data block
         # i'th entry of `indices` contains the index of `filter_params` with corresponding CTF params
         filter_params, indices = np.unique(
-            df[CTF_params].astype(self.dtype).values, return_inverse=True, axis=0
+            np.vstack(
```

Comment:
Think `np.stack` suffices here. Also maybe split call up into two lines (one to construct the raw params, one to get the unique indices).

Reply:
Refactored

Reply:
Thanks.